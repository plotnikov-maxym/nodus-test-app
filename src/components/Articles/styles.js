export const styles = ({spacing: {unit}, palette}) => ({
  root: {
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-start",
    padding: unit * 3,
  },
  flexRow: {
    display: "flex",
    flexDirection: "row",
    alignItems: "flex-end",
    justifyItems: "flex-start",
  },
  flexColumn: {
    display: "flex",
    flexDirection: "column",
    alignItems: "flex-start",
    padding: unit * 2,
    flexBasis: "90%",
  },
  card: {
    marginBottom: unit,
  },
  group: {
    margin: `${unit}px 0`,
    width: "100%",
  },
  date: {
    color: palette.grey.A200,
  },
  detailsBtn: {
    flexBasis: "10%",
    margin: unit * 2,
  },
});
