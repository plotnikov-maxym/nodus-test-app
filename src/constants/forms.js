export const ITEMS = [
  {
    value: 'arthroplasty',
    label: 'Total Hip Arthroplasty Revision',
  },
  {
    value: 'anterior',
    label: 'Anterior Lumbar Interbody Fusion Infection Rate',
  }];
