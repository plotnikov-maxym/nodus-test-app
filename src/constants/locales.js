export const locale = {
  WELCOME: "Welcome to Nodus research",
  SEARCH: "Search",
  SEARCH_IN_PUBMED: "Search in PubMed",
  TOO_SHORT_SEARCH_QUERY: "Please enter more than 3 characters for search.",
  ENTER_QUERY: "Please enter Your query to search.",
  NO_MATCHES: "No matches found.",
  FAVOURITE_ITEMS: "My favourite search items",
  DETAILS: "Details",
  ADD_TO_FAVOURITES: "Add to favourites",
  NO_FAVOURITES:
    "You don't have any favourites at the moment. You can add them at any time.",
};
