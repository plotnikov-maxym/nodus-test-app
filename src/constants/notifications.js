export const NOTIFICATION_TYPES = Object.freeze(["error", "info"]);
