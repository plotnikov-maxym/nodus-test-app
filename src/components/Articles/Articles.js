import React from "react";
import {array, func, object} from "prop-types";
import Typography from "@material-ui/core/Typography";
import Paper from "@material-ui/core/Paper";
import Button from "@material-ui/core/Button";
import {styles} from "./styles";
import {withStyles} from "@material-ui/core/styles";
import {locale} from "../../constants/locales";

export const Articles = ({classes, articles, fetchAbstract}) => {
  if (!articles.length) {
    return (
      <div className={classes.root}>
        <Typography variant="button" gutterBottom>
          {locale.NO_MATCHES}
        </Typography>
      </div>
    );
  }

  return (
    <div className={classes.root}>
      <div className={classes.group}>
        {articles.map((item, index) => {
          return (
            <Paper elevation={2} className={classes.card} key={item.pmid}>
              <div className={classes.flexRow}>
                <div className={classes.flexColumn}>
                  <Typography variant="h6" gutterBottom>
                    {item.title}
                  </Typography>
                  <Typography variant="subtitle1" display="inline">
                    {item.authors}
                  </Typography>
                  <Typography variant="subtitle2" display="inline">
                    {item.raw.source} ({item.raw.pubdate})
                  </Typography>
                </div>
                <Button
                  className={classes.detailsBtn}
                  variant="text"
                  color="primary"
                  onClick={() => fetchAbstract(item.pmid, index)}
                >
                  {locale.DETAILS}
                </Button>
              </div>
              {item.abstract && (
                <div className={classes.flexRow}>
                  <div className={classes.flexColumn}>
                    <Typography variant="subtitle1">{item.abstract}</Typography>
                  </div>
                </div>
              )}
            </Paper>
          );
        })}
      </div>
    </div>
  );
};

Articles.propTypes = {
  classes: object.isRequired,
  articles: array.isRequired,
  fetchAbstract: func.isRequired,
};

export default withStyles(styles)(Articles);
