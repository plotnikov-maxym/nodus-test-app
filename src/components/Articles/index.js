import {connect} from "react-redux";
import Articles from "./Articles";
import {fetchAbstract} from "../../actions/articles";

const mapStateToProps = state => ({
  articles: state.articles && state.articles.list,
});

const mapDispatchToProps = dispatch => ({
  fetchAbstract: (pmid, index) => dispatch(fetchAbstract(pmid, index)),
});
export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Articles);
