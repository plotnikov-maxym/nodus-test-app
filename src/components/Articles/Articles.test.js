import React from "react";
import renderer from "react-test-renderer";
import {extend} from "underscore";
import {createShallow} from "@material-ui/core/test-utils";
import {Articles} from "./Articles";
import Button from "@material-ui/core/Button";
import Paper from "@material-ui/core/Paper";
import Typography from "@material-ui/core/Typography";
import {locale} from "../../constants/locales";

describe("<Articles />", () => {
  let shallow;

  beforeAll(() => {
    shallow = createShallow();
  });

  describe("Articles component", () => {
    it("matches snapshot if no Articles provided", async () => {
      const wrapper = await renderComponent({articles: []});
      expect(wrapper).toMatchSnapshot();
    });

    it("matches snapshot with Articles provided", async () => {
      const wrapper = await renderComponent();
      expect(wrapper).toMatchSnapshot();
    });

    it("renders the correct number of items", async () => {
      const wrapper = await getComponent();
      const props = getMockProps();
      const itemsQty = props.articles.length;
      expect(wrapper.find(Paper).children()).toHaveLength(itemsQty);
    });

    it("renders fallback if no favourites provided", async () => {
      const articles = [];
      const wrapper = await getComponent({articles});
      const itemsQty = articles.length;
      expect(wrapper.find(Paper).children()).toHaveLength(itemsQty);
      const fallbackText = wrapper
        .find(Typography)
        .findWhere(text => text.prop("variant") === "button");
      expect(fallbackText.props().children).toEqual(locale.NO_MATCHES);
    });
  });

  it("renders Details button and calls fetchAbstract function", async () => {
    const fetchAbstract = jest.fn();
    const wrapper = await getComponent({fetchAbstract});
    const {articles} = getMockProps();
    const addButton = wrapper
      .find(Button)
      .findWhere(btn => btn.prop("variant") === "text");
    expect(addButton).toExist();
    addButton.simulate("click");
    await expect(fetchAbstract).toHaveBeenCalledWith(articles[0].pmid, 0);
  });

  const getComponent = props => {
    const parsedProps = extend(getMockProps(), props);

    return shallow(<Articles {...parsedProps} />);
  };

  const renderComponent = props => {
    const parsedProps = extend(getMockProps(), props);

    return renderer.create(<Articles {...parsedProps} />).toJSON();
  };

  const getMockProps = () => ({
    classes: {
      card: "Articles-card-198",
      date: "Articles-date-200",
      flexColumn: "Articles-flexColumn-197",
      flexRow: "Articles-flexRow-196",
      group: "Articles-group-199",
      root: "Articles-root-195",
    },
    articles: [
      {
        authors:
          "Blevins Peratikos M, Weeks HL, Pisansky AJB, Yong RJ, Stringer EA",
        doi: "10.1093/pm/pnz083",
        pmid: 31120529,
        pubDate: "2019/05/23 00:00",
        raw: {
          uid: "31120529",
          pubdate: "2019 May 23",
          epubdate: "2019 May 23",
          source: "Pain Med",
          authors: [],
        },
        title:
          "Effect of Preoperative Opioid Use on Adversplasty in the United States",
      },
    ],
    fetchAbstract: () => {},
  });
});
