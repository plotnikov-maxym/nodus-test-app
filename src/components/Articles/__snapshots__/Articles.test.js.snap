// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Articles /> Articles component matches snapshot if no Articles provided 1`] = `
<div
  className="Articles-root-195"
>
  <span
    className="MuiTypography-root-1 MuiTypography-button-12 MuiTypography-gutterBottom-28"
  >
    No matches found.
  </span>
</div>
`;

exports[`<Articles /> Articles component matches snapshot with Articles provided 1`] = `
<div
  className="Articles-root-195"
>
  <div
    className="Articles-group-199"
  >
    <div
      className="MuiPaper-root-37 MuiPaper-elevation2-41 MuiPaper-rounded-38 Articles-card-198"
    >
      <div
        className="Articles-flexRow-196"
      >
        <div
          className="Articles-flexColumn-197"
        >
          <h6
            className="MuiTypography-root-1 MuiTypography-h6-18 MuiTypography-gutterBottom-28"
          >
            Effect of Preoperative Opioid Use on Adversplasty in the United States
          </h6>
          <h6
            className="MuiTypography-root-1 MuiTypography-subtitle1-19"
            display="inline"
          >
            Blevins Peratikos M, Weeks HL, Pisansky AJB, Yong RJ, Stringer EA
          </h6>
          <h6
            className="MuiTypography-root-1 MuiTypography-subtitle2-20"
            display="inline"
          >
            Pain Med
             (
            2019 May 23
            )
          </h6>
        </div>
        <button
          className="MuiButtonBase-root-90 MuiButton-root-64 MuiButton-text-66 MuiButton-textPrimary-67 MuiButton-flat-69 MuiButton-flatPrimary-70"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex="0"
          type="button"
        >
          <span
            className="MuiButton-label-65"
          >
            Details
          </span>
          <span
            className="MuiTouchRipple-root-93"
          />
        </button>
      </div>
    </div>
  </div>
</div>
`;
